contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,320 @@
+        game.stake = stake;
+        depositStake(stake);
+
+        emit PlayerOneMoves(gameKey, msg.sender, stake, msg.value);
+    }
+
+    function movePlayerTwo(bytes32 gameKey, bytes32 moveHash) external payable whenNotPaused {
+
+        require(moveHash != bytes32(0), "Move hash is empty");
+        Game storage game = games[gameKey];
+        require(game.step == Steps.PLAYER_ONE_MOVED, INVALID_STEP_MSG);
+
+        uint256 expiryDate = block.timestamp.add(FORFEIT_WINDOW);
+        game.expiryDate = expiryDate;
+        game.playerTwo = msg.sender;
+        game.playerTwoMoveHash = moveHash;
+        game.step = Steps.PLAYER_TWO_MOVED;
+        depositStake(game.stake);
+
+        emit PlayerTwoMoves(gameKey, msg.sender, msg.value, moveHash, expiryDate);
+    }
+
+    function depositStake(uint256 stake) internal {
+
+        if (msg.value == stake) {
+            return;
+        }
+
+        balances[msg.sender] = balances[msg.sender].add(msg.value).sub(stake);
+    }
+
+    function revealPlayerOne(bytes32 secret, Moves playerOneMove) external whenNotPaused {
+
+        require(playerOneMove >= Moves.ROCK, INVALID_MOVE_MSG);
+        bytes32 gameKey = createPlayerOneMoveHash(msg.sender, secret, playerOneMove);
+        Game storage game = games[gameKey];
+
+        require(game.step == Steps.PLAYER_TWO_MOVED, INVALID_STEP_MSG);
+        game.step = Steps.PLAYER_ONE_REVEALED;
+
+        uint256 expiryDate = block.timestamp.add(FORFEIT_WINDOW);
+        game.expiryDate = expiryDate;
+        game.playerOneMove = playerOneMove;
+
+        emit PlayerOneReveals(gameKey, msg.sender, playerOneMove, expiryDate);
+    }
+
+    function revealPlayerTwo(bytes32 gameKey, bytes32 secret, Moves playerTwoMove) external whenNotPaused {
+
+        require(playerTwoMove >= Moves.ROCK, INVALID_MOVE_MSG);
+        Game storage game = games[gameKey];
+        require(game.step == Steps.PLAYER_ONE_REVEALED, INVALID_STEP_MSG);
+        address playerTwo = game.playerTwo;
+
+        //Validate move
+        bytes32 expectedMoveHash = createPlayerTwoMoveHash(msg.sender, gameKey, secret, playerTwoMove);
+        require(game.playerTwoMoveHash == expectedMoveHash, "Move and secret do not match");
+
+        emit PlayerTwoReveals(gameKey, msg.sender, playerTwoMove);
+
+        uint256 stake = game.stake;
+        address playerOne = game.playerOne;
+        address winner;
+        address loser;
+        GameResult outcome = resolveGame(game.playerOneMove, playerTwoMove);
+
+        if (outcome == GameResult.DRAW) {
+            resetGame(game);
+            balances[playerOne] = balances[playerOne].add(stake);
+            balances[playerTwo] = balances[playerTwo].add(stake);
+            DrawRefund(gameKey, playerOne, playerTwo, stake, stake);
+
+            return;
+
+        } else if (outcome == GameResult.LEFTWIN) {
+            winner = playerOne;
+            loser = playerTwo;
+        } else if (outcome == GameResult.RIGHTWIN) {
```

Comment:
A note about this long `if else`, some people (not me) say that it is better to have a single return point. Here you have 2: on line 199, and on line 215. I don't have an elegant solution for your case, only:

```solidity
if (outcome == GameResult.DRAW) {
    ...
    // not return
} else {
    if (outcome == GameResult.LEFTWIN) {
        ...
    } else if (outcome == GameResult.RIGHTWIN) { 
        ...
    } else {
        assert(false);
    }
    winner and loser handling
}
```

Reply:
I appreciate getting these perspectives. I need to go back to my Uncle Bob Martin books and refresh what I learnt. I probably should use the early return principle with more discretion and in the right context.

Reply:
I made the change you suggested and it looks much cleaner.